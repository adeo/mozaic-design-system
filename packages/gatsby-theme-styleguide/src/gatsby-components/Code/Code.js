import React, { useState } from 'react'
import Highlight, { defaultProps } from 'prism-react-renderer'
import copyToClipboard from '../../utils/copy-to-clipboard'
import IconLibrairy from '../IconLibrairy'
import styled from 'styled-components'

const CopyButton = styled.button`
  width: 1.2rem;
  height: 1.2rem;
  box-sizing: border-box;
  cursor: pointer;
  background: rgb(42, 39, 52);
  box-shadow: none;
  border: none;
  position: absolute;
  right: 1.1rem;
  top: 6px;
  outline: none;
  padding: 0.1rem;

  svg {
    overflow: visible;
  }
`

const LineNumbered = styled.div`
  min-height: 1em;

  &:before {
    display: inline-block;
    width: 3em;
    overflow: hidden;
    font-size: 0.8em;
  }
`

const CodeBlock = styled.code`
  max-width: 100%;
  overflow: auto;
  margin: 0 !important;
  white-space: pre-wrap;
  border: none !important;
`

const Wrapper = styled.div`
  position: relative;
`

const Pre = styled.pre`
  margin-bottom: 0 !important;
  margin-top: 0 !important;
  position: auto !important;
  padding: 1rem 1rem 0.4rem 1rem !important;
`

const copyToClipboardClick = (str, toogleCopy) => {
  copyToClipboard(str)
    .then(() => {
      toogleCopy(true)
      setTimeout(() => toogleCopy(false), 3000)
    })
    .catch(() => false)
}

const CodeHilight = ({ code, language, fullScreen, isOpen }) => (
  <Highlight
    {...defaultProps}
    code={code}
    language={language}
    theme={undefined}
  >
    {({ className, style, tokens, getLineProps, getTokenProps }) => (
      <Pre className={className}>
        <CodeBlock
          isOpen={isOpen}
          fullScreen={fullScreen}
          className={className}
          style={style}
        >
          {tokens.map((line, i) => (
            <LineNumbered ln={i} {...getLineProps({ line, key: i })}>
              {line.map((token, key) => (
                <span {...getTokenProps({ token, key })} />
              ))}
            </LineNumbered>
          ))}
        </CodeBlock>
      </Pre>
    )}
  </Highlight>
)

/*
 * MDX passes the code block as JSX
 * we un-wind it a bit to get the string content
 * but keep it extensible so it can be used with just children (string) and className
 */
const Code = ({
  children,
  className = children.props ? children.props.className : ``,
  isOpen,
  fullScreen,
}) => {
  const [copied, setCopied] = useState(false)
  const language = className
    ? className
        .split(`language-`)
        .pop()
        .toLowerCase()
    : ''

  const content =
    children.props && children.props.children
      ? children.props.children
      : children

  return (
    <Wrapper>
      {content && (
        <CodeHilight
          fullScreen={fullScreen}
          code={content}
          language={language}
          isOpen={isOpen}
          theme={undefined}
        />
      )}
      <CopyButton
        onClick={() => {
          copyToClipboardClick(content, setCopied)
        }}
      >
        {copied ? (
          <IconLibrairy name="checked" size="0.8rem" fill="#ececec" />
        ) : (
          <IconLibrairy name="copy" size="0.8rem" fill="#ececec" />
        )}
      </CopyButton>
    </Wrapper>
  )
}

export default Code
