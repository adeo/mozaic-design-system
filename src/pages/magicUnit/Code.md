---
title: 'Code'
order: 3
---

### What is the magic unit ?
