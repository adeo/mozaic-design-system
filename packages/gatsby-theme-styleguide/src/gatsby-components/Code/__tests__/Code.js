import React from 'react'
import renderer from 'react-test-renderer'
import Code from '../Code'

test('<Code />', () => {
  const code = `@import 'settings-tools/_all-settings';`
  const tree = renderer.create(<Code children={code} />).toJSON()
  expect(tree).toMatchSnapshot()
  expect(tree.children[0].children[0].type).toBe('code')
})
