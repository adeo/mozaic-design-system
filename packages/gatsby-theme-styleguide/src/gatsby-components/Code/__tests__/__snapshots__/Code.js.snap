// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Code /> 1`] = `
.c4 {
  width: 1.2rem;
  height: 1.2rem;
  box-sizing: border-box;
  cursor: pointer;
  background: rgb(42,39,52);
  box-shadow: none;
  border: none;
  position: absolute;
  right: 1.1rem;
  top: 6px;
  outline: none;
  padding: 0.1rem;
}

.c4 svg {
  overflow: visible;
}

.c3 {
  min-height: 1em;
}

.c3:before {
  display: inline-block;
  width: 3em;
  overflow: hidden;
  font-size: 0.8em;
}

.c2 {
  max-width: 100%;
  overflow: auto;
  margin: 0 !important;
  white-space: pre-wrap;
  border: none !important;
}

.c0 {
  position: relative;
}

.c1 {
  margin-bottom: 0 !important;
  margin-top: 0 !important;
  position: auto !important;
  padding: 1rem 1rem 0.4rem 1rem !important;
}

<div
  className="c0"
>
  <pre
    className="c1 prism-code language-"
  >
    <code
      className="c2 prism-code language-"
      style={Object {}}
    >
      <div
        className="c3 token-line"
      >
        <span
          className="token plain"
        >
          @import 'settings-tools/_all-settings';
        </span>
      </div>
    </code>
  </pre>
  <button
    className="c4"
    onClick={[Function]}
  >
    <svg
      fill="#ececec"
      height="0.8rem"
      style={Object {}}
      viewBox="0 0 16 16"
      width="0.8rem"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M10 4v-4h-7l-3 3v9h6v4h10v-12h-6zM3 1.414v1.586h-1.586l1.586-1.586zM1 11v-7h3v-3h5v3l-3 3v4h-5zM9 5.414v1.586h-1.586l1.586-1.586zM15 15h-8v-7h3v-3h5v10z"
        fill="#ececec"
      />
    </svg>
  </button>
</div>
`;
